aspnetcore/blazor/fundamentals/render-modes.md

```diff
@@ -0,0 +1,295 @@
+
+`SharedMessage.razor`:
+
+```razor
+<button @onclick="UpdateMessage">Click me</button> @message
+
+@code {
+    private string message = "Not clicked yet.";
+
+    private void UpdateMessage()
+    {
+        message = "Somebody clicked me!";
+    }
+}
+```
+
+If the `SharedMessage` component is placed in a statically-rendered parent component, the `SharedMessage` component is also rendered statically and isn't interactive. The button doesn't call `UpdateMessage`, and the message isn't updated.
+
+`PassedRenderMode1.razor`:
+
+```razor
+@page "/passed-render-mode-1"
+
+<SharedMessage />
+```
+
+If the `SharedMessage` component is placed in a server-rendered component, it adopts server-side rendering. The `SharedMessage` component is interactive over a SignalR connection to the client. The button calls `UpdateMessage`, and the message is updated.
+
+`PassedRenderMode2.razor`:
+
+```razor
+@page "/passed-render-mode-2"
+@attribute [RenderModeServer]
+
+<SharedMessage />
+```
+
+If the `SharedMessage` component is placed in a client-rendered component, it adopts client-side rendering. The `SharedMessage` component is interactive on the client with the .NET WebAssembly-based runtime. The button calls `UpdateMessage`, and the message is updated.
+
+`PassedRenderMode3.razor`:
+
+```razor
+@page "/passed-render-mode-3"
+@attribute [RenderModeWebAssembly]
+
+<SharedMessage />
+```
+
+If the `SharedMessage` component is placed in a dynamically-rendered component, where the render mode is determined at runtime, it adopts server-side rendering initially, then client-side rendering after the component and Blazor bundle are downloaded and the .NET runtime activates on the client. The component is interactive. The button calls `UpdateMessage`, and the message is updated.
+
+`PassedRenderMode4.razor`:
+
+```razor
+@page "/passed-render-mode-4"
+@attribute [RenderModeAuto]
+
+<SharedMessage />
+```
+
+## Set the render mode for the entire app
+
+To set the render mode for the entire app, indicate the render mode at the highest level component in the app's component hierarchy, typically the `Routes` component (`Components/Routes.razor`) for apps based on the Blazor Web App project template:
+
+```razor
+@attribute {RENDER MODE}
+
+<Router AppAssembly="@typeof(App).Assembly" ...>
+    <Found Context="routeData">
+        <RouteView RouteData="@routeData" DefaultLayout="@typeof(Layout.MainLayout)" />
+        <FocusOnNavigate RouteData="@routeData" Selector="h1" />
+    </Found>
+</Router>
+```
+
+In the preceding example, the `{RENDER MODE}` placeholder is the render mode (`[RenderModeServer]`, `[RenderModeWebAssembly]`, or `[RenderModeAuto]`). The render mode propagates down the component hierarchy to all of the components in the app.
+
+## API extensions that support components in Blazor Web Apps
```

Comment:
```suggestion
## Enable support for interactive render modes
```

Reply:
Should this section go earlier in the doc? The above samples won't work if you haven't enabled the appropriate render modes.